### PR TITLE
HPCC-31989 Implement new streaming input and output classes for spills

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith this is an initial version.  Believed working - it includes unit tests for reading and writing in parallel with and without compression and different buffer sizes.
I still have a long list of changes/improvements to make, but it should at least be enough to see if it provides what you need.
Main files to look at to understand its use are jstream.hpp and runParallelStream() test within jstreamtests.cpp

File: system/jlib/jlzw.cpp [hunk 1]
```diff
@@ -463,6 +463,16 @@ size32_t CExpanderBase::expandNext(MemoryBuffer & target)
     return 0;
 }
 
+size32_t CExpanderBase::expandDirect(size32_t destSize, void * dest, size32_t srcSize, const void * src)
+{
+    return 0;
```
Comment (jakesmith) [inline system/jlib/jlzw.cpp:468]: throwUnimplemented?

File: system/jlib/jlzw.cpp [hunk 2]
```diff
@@ -1723,6 +1738,11 @@ class jlib_decl CRandRDiffCompressor : public ICompressor, public CInterface
 
     virtual size32_t compressBlock(size32_t destSize, void * dest, size32_t srcSize, const void * src) override { return 0; }
 
+    virtual size32_t compressDirect(size32_t destSize, void * dest, size32_t srcSize, const void * src, size32_t * numCompressed) override
+    {
+        return 0;
```
Comment (jakesmith) [inline system/jlib/jlzw.cpp:1743]: throwUnimplemented

File: system/jlib/jstream.hpp [hunk 3]
```diff
@@ -43,4 +43,90 @@ extern jlib_decl IByteInputStream *createInputStream(StringBuffer &from);
...
+
+static constexpr size32_t BufferTooSmall = (size32_t)-1;
+static constexpr offset_t UnknownOffset = (offset_t)-1;
+interface ISerialInputStream: extends IInterface
+{
+    virtual size32_t read(size32_t len, void * ptr) = 0;            // returns size read, result < len does NOT imply end of file
+    virtual void skip(size32_t sz) = 0;
+    virtual void get(size32_t len, void * ptr) = 0;                 // exception if no data available
+    virtual bool eos() = 0;                                         // no more data
+    virtual void reset(offset_t _offset,offset_t _flen) = 0;       // input stream has changed - restart reading
```
Comment (jakesmith) [inline system/jlib/jstream.hpp:56]: indentation of comment

File: system/jlib/jstream.hpp [hunk 4]
```diff
@@ -43,4 +43,90 @@ extern jlib_decl IByteInputStream *createInputStream(StringBuffer &from);
...
+*/
+
+interface ISerialOutputStream: extends IInterface
+{
+    virtual size32_t write(size32_t len, const void * ptr) = 0; // returns size written
+    virtual void put(size32_t len, const void * ptr) = 0;       // throws an error if cannot write the full size.
+    virtual void flush() = 0;
+};
+
+interface IBufferedSerialOutputStream: extends ISerialOutputStream
```
Comment (jakesmith) [inline system/jlib/jstream.hpp:97]: spacing before : extends and on line 90

File: system/jlib/jstream.hpp [hunk 5]
```diff
@@ -43,4 +43,90 @@ extern jlib_decl IByteInputStream *createInputStream(StringBuffer &from);
...
+{
+    virtual byte * reserve(size32_t wanted, size32_t & got) = 0;    // get a pointer to a contiguous block of memory to write to.
+    virtual void commit(size32_t written)= 0 ;      // commit the data written to the block returned by reserve
```
Comment (jakesmith) [inline system/jlib/jstream.hpp:100]: space before =

File: system/jlib/jstream.cpp [hunk 6]
```diff
@@ -282,3 +283,944 @@ IByteInputStream *createInputStream(int handle)
...
+//===========================================================================
+
+// This class ensures that data is read in fixed block sizes - NOT a fixed buffer size.
+// This means the buffer size is likely to be bigger than the block size - the class is passed
+// an initial estimate for the potential overlap.
+
+class CBlockedSerialInputStream : public CInterfaceOf<IBufferedSerialInputStream>
+{
+public:
+    CBlockedSerialInputStream(ISerialInputStream * _input, size32_t _blockReadSize, size32_t expectedPeekSize)
```
Comment (jakesmith) [inline system/jlib/jstream.cpp:296]: it is not clear (enough) how expectedPeekSize will be used, bettert to keep internal?

File: system/jlib/jstream.cpp [hunk 7]
```diff
@@ -282,3 +283,944 @@ IByteInputStream *createInputStream(int handle)
...
+            //While there are blocks larger than the buffer size read directly into the target buffer
+            while (sizeRead + blockReadSize <= len)
+            {
+                size32_t got = readNextBlock(blockReadSize, target+sizeRead);
+                if ((got == 0) || (got == BufferTooSmall))
+                    break;
+                sizeRead += got;
+                nextBlockOffset += got;
+            }
+
```
Comment (jakesmith) [inline system/jlib/jstream.cpp:327]: extra newline

File: system/jlib/jstream.cpp [hunk 8]
```diff
@@ -282,3 +283,944 @@ IByteInputStream *createInputStream(int handle)
...
+        {
+            size32_t toCopy = std::min(len, available());
+            memcpy(target, data(bufferOffset), toCopy);
+            bufferOffset += toCopy;
+            sizeRead += toCopy;
+            if (likely(sizeRead == len))
+                return sizeRead;
+        }
+
+        if (directRead)
```
Comment (jakesmith) [inline system/jlib/jstream.cpp:316]: why not unconditional?

File: system/jlib/jstream.cpp [hunk 9]
```diff
@@ -282,3 +283,944 @@ IByteInputStream *createInputStream(int handle)
...
+        {
+            if (endOfStream)
+                break;
+            readNextBlock(); // will be appended onto the end of the existing buffer
+        }
+        got = available();
+        return data(bufferOffset);
+    }
+
+
```
Comment (jakesmith) [inline system/jlib/jstream.cpp:368]: double newline

File: system/jlib/jstream.cpp [hunk 10]
```diff
@@ -282,3 +283,944 @@ IByteInputStream *createInputStream(int handle)
...
+            return len;
+        }
+        return slowWrite(len, ptr);
+    }
+
+    size32_t slowWrite(size32_t len, const void * ptr) __attribute__((noinline))
+    {
+        size32_t sizeWritten = 0;
+        const byte * src = (const byte *)ptr;
+        if (likely(!isSuspended()))
```
Comment (jakesmith) [inline system/jlib/jstream.cpp:708]: Add comment to explain approach.

File: system/jlib/jstream.cpp [hunk 11]
```diff
@@ -282,3 +283,944 @@ IByteInputStream *createInputStream(int handle)
...
+        if (likely(!isSuspended() && len <= space && blockWriteSize != 0))
+        {
+            memcpy(data(bufferOffset), ptr, len);
+            bufferOffset += len;
+            return len;
+        }
+        return slowWrite(len, ptr);
+    }
+
+    size32_t slowWrite(size32_t len, const void * ptr) __attribute__((noinline))
```
Comment (jakesmith) [inline system/jlib/jstream.cpp:704]: splitWrite ? slow is a little unintuitive

**Review by jakesmith — APPROVED** [review]
can't spot any real issues, looks good (a few more minor comments/questions)

**Review by jakesmith — APPROVED** [review]
@ghalliday - I can't see any issues in the new commits.

**Pull request merged**